kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java

```diff
@@ -43,12 +45,49 @@ public class KafkaProxyFrontendHandler extends ChannelInboundHandlerAdapter {
 
     private static final Logger LOGGER = LogManager.getLogger(KafkaProxyFrontendHandler.class);
 
+    /**
+     * Builds a request pipeline for incoming requests from the downstream client.
+     * @param filters The filters in the pipeline (response filters won't be added to the result).
+     * @return A list of channel handlers
+     */
+    static List<ChannelHandler> buildRequestPipeline(List<KrpcFilter> filters) {
+        // Note: we could equally use a single ChannelInboundHandler which itself dispatched to each filter.
+        // Using a ChannelInboundHandler-per-filter model means that we're not occupying the CPU for the
+        // whole filterchain execution => higher latency, but higher throughput.
+        List<ChannelHandler> requestFilterHandlers = new ArrayList<>(filters.size());
+        for (var filter : filters) {
+            if (filter instanceof KrpcRequestFilter) {
+                requestFilterHandlers.add(new SingleRequestFilterHandler((KrpcRequestFilter) filter));
+            }
+        }
+        return requestFilterHandlers;
+    }
+
+    /**
+     * Builds a response pipeline for incomping responses from the upstream server.
+     * @param filters The filters in the pipeline (request filters won't be added to the result).
+     * @return A list of channel handlers
+     */
+    static List<ChannelHandler> buildResponsePipeline(List<KrpcFilter> filters) {
+        // Note: we could equally use a single ChannelInboundHandler which itself dispatched to each filter.
+        // Using a ChannelInboundHandler-per-filter model means that we're not occupying the CPU for the
+        // whole filterchain execution => higher latency, but higher throughput.
+        List<ChannelHandler> responseFilterHandlers = new ArrayList<>(filters.size());
+        for (var filter : filters) {
+            if (filter instanceof KrpcResponseFilter) {
+                responseFilterHandlers.add(new SingleResponseFilterHandler((KrpcResponseFilter) filter));
+            }
+        }
+        return responseFilterHandlers;
+    }
+
     private final String remoteHost;
     private final int remotePort;
     private final Map<Integer, Correlation> correlation;
     private final boolean logNetwork;
     private final boolean logFrames;
-    private final List<ChannelHandler> responseFilterHandlers;
+    private final List<KrpcFilter> filters;
```

Comment:
Given that filters are not supposed to change here (I suppose), we can just use a simple array here and save external iterations to create (and allocate) and iterator while looping over it